man/man3/Session.3
.TH "Session" 3 "Fri Oct 7 2011" "Version 1.0" "Launchpad" \" -*- nroff -*-
.ad l
.nh
.SH NAME
Session \- 
.SH SYNOPSIS
.br
.PP
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fB__set\fP ($key, $value)"
.br
.ti -1c
.RI "\fB__get\fP ($key)"
.br
.ti -1c
.RI "\fB__isset\fP ($key)"
.br
.ti -1c
.RI "\fB__unset\fP ($key)"
.br
.ti -1c
.RI "\fBis_logged_in\fP ()"
.br
.ti -1c
.RI "\fBlogin\fP ($user_id)"
.br
.ti -1c
.RI "\fBlogout\fP ()"
.br
.ti -1c
.RI "\fBmessage\fP ($msg='')"
.br
.ti -1c
.RI "\fBopen\fP ()"
.br
.ti -1c
.RI "\fBclose\fP ()"
.br
.ti -1c
.RI "\fBread\fP ($id)"
.br
.ti -1c
.RI "\fBwrite\fP ($id, $data)"
.br
.ti -1c
.RI "\fBdestroy\fP ($id)"
.br
.ti -1c
.RI "\fBclean\fP ($timeout)"
.br
.in -1c
.SH "Detailed Description"
.PP 
A class for the creation and acessing of session variables 
.SH "Member Function Documentation"
.PP 
.SS "Session::__get ($key)"Gets a sessin variable. Use: $session->foo;
.PP
\fBParameters:\fP
.RS 4
\fI$key\fP - the session to retrieve 
.RE
.PP
\fBReturns:\fP
.RS 4
string 
.RE
.PP

.SS "Session::__isset ($key)"check whether a session variable is set. Use: isset($session->foo);
.PP
\fBParameters:\fP
.RS 4
\fI$key\fP - The variable to check 
.RE
.PP
\fBReturns:\fP
.RS 4
boolean 
.RE
.PP

.SS "Session::__set ($key, $value)"sets a session veriable. Use: $session->foo = bar;
.PP
\fBParameters:\fP
.RS 4
\fI$key\fP - the variable name 
.br
\fI$value\fP - the variable value 
.RE
.PP
\fBReturns:\fP
.RS 4
null 
.RE
.PP

.SS "Session::__unset ($key)"unsets a session variable is set. Use: unset($session->foo);
.PP
\fBParameters:\fP
.RS 4
\fI$key\fP - The variable to unset 
.RE
.PP
\fBReturns:\fP
.RS 4
null 
.RE
.PP

.SS "Session::clean ($timeout)"Deletes expired sessions
.PP
\fBParameters:\fP
.RS 4
\fI$timeout\fP - Unix timestamp representing the end of the session life 
.RE
.PP
\fBReturns:\fP
.RS 4
boolean 
.RE
.PP

.SS "Session::close ()"Closes database connection
.PP
\fBReturns:\fP
.RS 4
boolean 
.RE
.PP

.SS "Session::destroy ($id)"Removes a session from the database
.PP
\fBParameters:\fP
.RS 4
\fI$id\fP - description 
.RE
.PP
\fBReturns:\fP
.RS 4
boolean - success 
.RE
.PP

.SS "Session::is_logged_in ()"Check whether the current user is logged in
.PP
\fBReturns:\fP
.RS 4
booelan 
.RE
.PP

.SS "Session::login ($user_id)"Logins in a user
.PP
\fBParameters:\fP
.RS 4
\fI$user_id\fP - The user id of the user to log in. 
.RE
.PP
\fBReturns:\fP
.RS 4
null 
.RE
.PP

.SS "Session::logout ()"Logs out the currently logged in user
.PP
\fBReturns:\fP
.RS 4
boolean 
.RE
.PP

.SS "Session::message ($msg = \fC''\fP)"Set the message session variable.
.PP
\fBParameters:\fP
.RS 4
\fI$msg\fP - The message to set. Pass null or empty string to unset the current message; 
.RE
.PP
\fBReturns:\fP
.RS 4
boolean | string 
.RE
.PP

.SS "Session::open ()"Opens a new database connection for string session
.PP
\fBReturns:\fP
.RS 4
boolean 
.RE
.PP

.SS "Session::read ($id)"Reads the session from the database
.PP
\fBParameters:\fP
.RS 4
\fI$id\fP - The id of the session 
.RE
.PP
\fBReturns:\fP
.RS 4
array - \fBSession\fP data 
.RE
.PP

.SS "Session::write ($id, $data)"Writes session data to the database
.PP
\fBParameters:\fP
.RS 4
\fI$id\fP - The id of the session 
.br
\fI$data\fP - The data to be stored 
.RE
.PP
\fBReturns:\fP
.RS 4
boolean - success of execution 
.RE
.PP


.SH "Author"
.PP 
Generated automatically by Doxygen for Launchpad from the source code.


lib/session.class.php
<?php

/**
 * @author Jonah Werre <jonahwerre@gmail.com>
 * @version 1.0
 * @copyright Jonah Werre <jonahwerre@gmail.com>, 27 June, 2011
 * @package default
 **/

/**
 * A class for the creation and acessing of session variables 
 */      

class Session
{
	
	private $logged_in = false;
	private $database;
	
	function __construct()
	{	

		if (USE_DB_SESSION) {
			session_set_save_handler( array( &$this, "open" ), array( &$this, "close" ), array( &$this, "read" ), array( &$this, "write"), array( &$this, "destroy"), array( &$this, "clean" ) );
		}
		
		session_start();

		if( $this->check_login() ) {
			$this->active();
		} else {
			
	    }
	}
	
	/**
	 * sets a session veriable. Use: $session->foo = bar;
	 *
	 * @param string $key - the variable name
	 * @param string $value - the variable value
	 * @return null
	 **/
	public function __set($key, $value)
	{
        $_SESSION[$key] = $value;
    }

	/**
	 * Gets a sessin variable. Use: $session->foo;
	 *
	 * @param string $key - the session to retrieve 
	 * @return string
	 **/
	public function __get($key)
	{
        if (array_key_exists($key, $_SESSION)) {
            return $_SESSION[$key];
		}else{
			return false;
		}
    }

	/**
	 * check whether a session variable is set. Use: isset($session->foo);
	 *
	 * @param string $key - The variable to check
	 * @return boolean
	 **/
    public function __isset($key) {
        return isset($_SESSION[$key]);
    }

	/**
	 * unsets a session variable is set. Use: unset($session->foo);
	 *
	 * @param string $key - The variable to unset
	 * @return null
	 **/
    public function __unset($key) {
        unset($_SESSION[$key]);
    }

	/**
	 * Check whether the current user is logged in
	 *
	 * @return booelan
	 **/
	public function is_logged_in()
	{
		return $this->logged_in;
	}
	/**
	 * Logins in a user
	 *
	 * @param string $user_id - The user id of the user to log in.
	 * @return null
	 **/
	public function login($user_id)
	{
		if($this->logged_in){
			$this->logout();
		}
		if($user_id){
			$this->user_id = $user_id;
			$this->logged_in = true;
		}
	}
	/**
	 * Logs out the currently logged in user
	 *
	 * @return boolean
	 **/
	public function logout()
	{
		session_destroy();
		session_unset();
		$this->logged_in = false;
	}
	
	/**
	* Set the message session variable.
	*
	* @param string $msg - The message to set. Pass null or empty string to unset the current message;
	* @return boolean | string
	*/
	public function message($msg="")
	{
		if (!empty($msg)) {
			$_SESSION['message'] = $msg;
		}else {
			$msg = ( isset($_SESSION['message']) ) ? $_SESSION['message'] : "";
			unset($_SESSION['message']);
		}
		return $msg;
	}
	
	
	//--------------------------------------
	//  PRIVATE & PROTECTED METHODS
	//--------------------------------------
	
	/**
	 * Checks if a logged in user still has an active session. If not, logs the user out.
	 *
	 * @return null
	 **/
	private function active()
	{
		if (isset($_SESSION['last_active']) && (time() - $_SESSION['last_active'] > SESSION_LIFE)) {
			$this->logout();
		}
		$_SESSION['last_active'] = time(); // update last activity time stamp

		// Avoid attacks on sessions by regenerating session id every 30 minutes
		if (!isset($_SESSION['created'])) {
			$_SESSION['created'] = time();
		} else if (time() - $_SESSION['created'] > SESSION_LIFE) {
			session_regenerate_id(true);    // change session ID for the current session and invalidate old session ID
			$_SESSION['created'] = time();  // update creation time
		}
			
	}
	/**
	 * Checks if 'user_id' in session is logged in;
	 *
	 * @return null
	 **/
	private function check_login()
	{
		if( isset($_SESSION['user_id']) ){
			$this->logged_in = true;
		}else{
			unset($this->user_id);
			$this->logged_in = false;
		}
	}
	
	
	//******************************************
	// OVERRDING PHP SESSION FUNCTIONS
	// store session in database instead of file. In confi.php set USE_DB_SESSION to true;
	//******************************************
	/**
	 * Opens a new database connection for string session
	 *
	 * @return boolean
	 **/
	public function open() {
		$this->database = new DatabasePDO();
        return true;
    }
	/**
	 * Closes database connection
	 *
	 * @return boolean
	 **/
    public function close() {
		//$this->database->close_connection();
		//$this->database = NULL;
        return true;
    }
	/**
	 * Reads the session from the database
	 *
	 * @param int $id - The id of the session
	 * @return array - Session data
	 **/
	public function read( $id ) {
		$sql = "SELECT session_data FROM sessions WHERE id = '$id' AND expires > ". time();
		try{
			$this->database->query($sql);
			$result = $this->database->fetch_array();
		}catch(PDOException $e){
			echo "This is the problem: ". $e->getMessage();
		}
		return $result['session_data'];
	}
	
	/**
	 * Writes session data to the database
	 *
	 * @param int $id - The id of the session
	 * @param string $data - The data to be stored
	 * @return boolean - success of execution
	 **/
	public function write( $id, $data ) {
		$time = time() + SESSION_LIFE;
		$sql = "INSERT INTO sessions (id, session_data, expires) VALUES ('$id',?, $time)";
		try{
			return $this->database->execute($sql, array($data));
		} catch(PDOException $e) { 
			if($e->getCode() == 23000) { // duplicate entry
				$sql = "UPDATE sessions SET session_data = ?, expires = $time WHERE id = '$id'";
				return $this->database->execute($sql, array($data) ); 
			}
		}
	}

	/**
	 * Removes a session from the database
	 *
	 * @param int $id - description
	 * @return boolean - success
	 **/
	public function destroy( $id ) {
		$sql = "DELETE FROM sessions WHERE id = '$id'";
		return $this->database->execute($sql);
	}
	/**
	 * Deletes expired sessions
	 *
	 * @param number $timeout - Unix timestamp representing the end of the session life
	 * @return boolean
	 **/
	public function clean($timeout) {
		$max_time = time() - $timeout;
		$sql = "DELETE FROM sessions WHERE expires < ?";
		return $this->database->execute($sql, array($max_time));
	}
	
}

$session = new Session();
?>
